### PR TITLE
[SPARK-23377][ML] Fixes Bucketizer with multiple columns persistence bug

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Since 2.3, `Bucketizer` supports multiple input/output columns. We will check if exclusive params are set during transformation. E.g., if `inputCols` and `outputCol` are both set, an error will be thrown.

However, when we write `Bucketizer`, looks like the default params and user-supplied params are merged during writing. All saved params are loaded back and set to created model instance. So the default `outputCol` param in `HasOutputCol` trait will be set in `paramMap` and become an user-supplied param. That makes the check of exclusive params failed.

This patch changes `DefaultParamsWriter` and only save user-supplied params.

The multi-column `QuantileDiscretizer` also has the same issue.

## How was this patch tested?

Modified test.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 1]
```diff
@@ -864,6 +864,11 @@ trait Params extends Identifiable with Serializable {
     extractParamMap(ParamMap.empty)
   }
 
+  /**
+   * Extracts the user-supplied params.
+   */
+  final def extractUserParamMap(): ParamMap = paramMap
```
Comment (mgaido91) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: can't we just make `paramMap` `private[ml]`?

**Review by mgaido91 — COMMENTED** [review]
I am just wondering whether we should persist the default params too (in case they are changed across multiple versions) but in a separate section. WDYT?

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala (hunk 1 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: In this way I think you can also avoid the MiMa failure...

**Comment by mgaido91** [comment]
@viirya that's a good question. Honestly my idea is that if the user doesn't set a value, he/she doesn't care about it, so it is good to use the new version default IMHO. But it is also true that changing a default may cause unexpected behavior in user code.

So, it LGTM, but I'd like to hear others' opinion on this too.

**Comment by viirya** [comment]
Yeah, IMHO, when the user loads a model from old version into new version to run, I think it is reasonable to run it with current default value because the param is not explicitly set and should use "default" value of current system.

Thanks for your comment. Let's wait for others' option.

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala (hunk 1 above)
Comment (viirya) [reply mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: Looks like it still can't avoid the MiMa failure.

**Comment by MrBago** [comment]
I believe this will break persistence for LogisticRegression. I believe the issue is that the `threshold` param on LogisticRegressionModel doesn't get a default directly, but only gets it during the call to `fit` on LogisticRegression. This is currently fine because the Model can only be created by fitting or by being read from disk and in both case some value gets set for threshold. With this change that's no longer the case. Here's a test to confirm, https://github.com/apache/spark/commit/5db2108224accdf848b41ef0d8d1c312b49f49c6.

I believe LinearRegression may have a similar issue.

Our current tests don't seem to cover this kind of thing so I think we should improve test coverage if we want to make this kind of change.

**Comment by viirya** [comment]
Not only `threshold`, the default params of `NaiveBayes`, `LogisticRegression` (maybe more, I'm looking up now) are all set in the estimator, not in their model. The models are received the default values at the end of `fit`.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
cc @MLnick @jkbradley

**Comment by jkbradley** [comment]
Thanks for the patch @viirya 
As always, I'll request that we put design decisions & long discussions in JIRA so that they are easier to uncover.  It can also be good to get quick feedback about design before implementation.  I'll comment in JIRA.

**Comment by viirya** [comment]
@jkbradley Thanks! I will post the problem and proposed design on the JIRA.

**Comment by viirya** [comment]
I'd close this and favor the quick fix #20594 based on the discussion in JIRA. Will re-open it if it is needed later.

**Pull request closed without merging**